library/std/arithmetic_internal.qs

```diff
@@ -0,0 +1,163 @@
+    @Config(Full)
+    internal operation FullAdderInc(carryIn : Qubit, x : Qubit, y : Qubit, carryOut : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            // TODO: cannot use `Carry` operation here
+            CNOT(carryIn, x);
+            CNOT(carryIn, y);
+            CCNOT(x, y, carryOut);
+            CNOT(carryIn, carryOut);
+            CNOT(carryIn, x);
+            CNOT(x, y);
+        }
+        adjoint auto;
+
+        controlled (ctls, ...) {
+            Fact(Length(ctls) == 1, "FullAdder should be controlled by exactly one control qubit.");
+
+            let ctl = ctls[0];
+            use helper = Qubit();
+
+            Carry(carryIn, x, y, helper);
+            CCNOT(ctl, helper, carryOut);
+            Controlled Uncarry(ctls, (carryIn, x, y, helper));
+        }
+        controlled adjoint auto;
+    }
+
+    // Computes carryOut := carryIn + x + y
+    @Config(Full)
+    internal operation FullAdder(carryIn : Qubit, x : Qubit, y : Qubit, carryOut : Qubit) : Unit is Adj {
+        CNOT(x, y);
+        CNOT(x, carryIn);
+        ApplyAndWith0Target(y, carryIn, carryOut);
+        CNOT(x, y);
+        CNOT(x, carryOut);
+        CNOT(y, carryIn);
+    }
+
+    /// # Summary
+    /// Compute carry bit for a full adder.
+    @Config(Full)
+    internal operation Carry(carryIn : Qubit, x : Qubit, y : Qubit, carryOut : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            CNOT(carryIn, x);
+            CNOT(carryIn, y);
+            ApplyAndWith0Target(x, y, carryOut);
+            CNOT(carryIn, carryOut);
+        }
+        adjoint auto;
+        controlled (ctls, ...) {
+            // TODO: Is it not controlled actually?
+            Carry(carryIn, x, y, carryOut);
+        }
+        controlled adjoint auto;
+    }
+
+    /// # Summary
+    /// Uncompute carry bit for a full adder.
+    @Config(Full)
+    internal operation Uncarry(carryIn : Qubit, x : Qubit, y : Qubit, carryOut : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            CNOT(carryIn, carryOut);
+            Adjoint ApplyAndWith0Target(x, y, carryOut);
+            CNOT(carryIn, x);
+            CNOT(x, y);
+        }
+        adjoint auto;
+        controlled (ctls, ...) {
+            Fact(Length(ctls) == 1, "Uncarry should be controlled by exactly one control qubit.");
+
+            let ctl = ctls[0];
+
+            CNOT(carryIn, carryOut);
+            Adjoint ApplyAndWith0Target(x, y, carryOut);
+            CCNOT(ctl, x, y); // Controlled X(ctls + [x], y);
+            CNOT(carryIn, x);
+            CNOT(carryIn, y);
+        }
+        controlled adjoint auto;
+    }
+
+    /// # Summary
+    /// Applies AND gate between `control1` and `control2` and stores the result
+    /// in `target` assuming `target` is in |0> state. This allows for an
+    /// otimized adjoint implementation.
+    @Config(Full)
+    internal operation ApplyAndWith0Target(control1 : Qubit, control2 : Qubit, target: Qubit) : Unit is Adj {
```

Comment:
This should eventually be a public operation (CC @swernli).

Reply:
I named it this way, but don't like this name much... It kind of sounds like you apply AND, and one of the arguments is "0 target", which is not what this operation does.

Reply:
It would be great to have this as an intrinsic operation (like `CCNOT`) and then have a simple name such as `And`.

Reply:
@msoeken Agreed. I think we want that to be the case eventually, and use that intrinsic both here and in the decompositions invoked in internal.qs (today called `PhaseCCX`). I'll include mention of this in our upcoming specifications so that it can be a clear part of the roadmap.